src/auth.ts

```diff
@@ -1,5 +1,182 @@
+      if (state.stateStore) {
+        const providedState = await state.stateStore.get();
+        await state.stateStore.del();
+        if (receivedState !== providedState) throw errors.stateMismatch;
+      }
+
+      // Remove `code` and `state` query parameters from url.
+      url.searchParams.delete("code");
+      url.searchParams.delete("state");
+      const redirectUrl = url.href;
+      history.pushState({}, "", redirectUrl);
+
+      // Exchange `code` search parameter for an OAuth authentication.
+      const route = state.serviceHost + state.pathPrefix + "/token";
+      const { data } = await state.request(route, {
+        method: "POST",
+        code,
+        redirectUrl,
+        state: receivedState,
+        // TODO: I’m not sure why `oauth-app.js` asks the `state`. GitHub API
+        // does not accept it as documented: https://docs.github.com/en/developers/apps/building-oauth-apps/authorizing-oauth-apps#parameters-1.
+        // This should be unnecessary: https://github.com/octokit/oauth-methods.js/blob/main/src/exchange-web-flow-code.ts#L78
+      });
+
+      // Cache and return the authentication.
+      const authentication = {
+        // TODO: `oauth-app.js` should be upgrade to be fully compatible. The
+        // middleware should return an authentcation with correct type. Shared
+        // types between `auth-oauth-user-client.js` and `oauth-app.js` may be
+        // refactored into @octokit/types. Below error-prone object assemblage
+        // may be completely removed.
+        type: "token",
+        tokenType: "oauth",
+        clientType: state.clientType,
+        clientId: state.clientId,
+        ...data,
+      };
+      state.authentication = authentication;
+      if (state.authStore) state.authStore.set(authentication);
+      return authentication;
+    }
+
+    // Redirect to web flow authorization url.
+    case "signIn": {
+      const oauthState = Math.random().toString(36).substr(2);
+      if (state.stateStore) state.stateStore.set(oauthState);
+      const getUrlOptions = {
+        clientType: state.clientType,
+        clientId: state.clientId,
+        request: state.request,
+        redirectUrl: location.href,
+        state: oauthState,
+        login: options.login,
+        allowSignup: options.allowSignup,
+      } as GetWebFlowAuthorizationUrlOptions<Client>;
+
+      // Code in this section is really ugly.
+      if (state.clientType === "oauth-app") {
+        (getUrlOptions as OAuthAppOptions).scopes = (
+          state as unknown as State<OAuthApp, Expiration>
+        ).scopes;
+      }
+      location.href = oauthAuthorizationUrl(getUrlOptions as any).url;
+      return null;
+    }
+
+    case "checkToken": {
+      // TODO: Should this request be issued against GitHub API instead of
+      // backend service? The `data` type is of type `OAuthMethods.ResetToken{OAuth|GitHub}AppResponse`
+      // which has nothing to do with `Authentication` here.
```

Comment:
Unfortunately GitHub's OAuth APIs are not part of `https://api.github.com`, and they are not CORS enabled. So we have to proxy them through our server in order to make them work for browsers